### PR TITLE
[aptos-rosetta] Make block info API, and get reconciliation working for Rosetta

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
To get Rosetta working, I had to reconstruct the blocks from ledger versions.  This currently doesn't handle any proper on disk storage, which is another step on this.  But, it does work entirely in memory if there is continually blocks being searched for.

This does a scan for the BlockMetadata transaction, and for the StateCheckpoint to figure out the boundaries for the block, and exposes this as an API with the block hash and other information needed.

There are a ton of performance things that could be updated, but this is an MVP.  Still need to work through the testing for the transfer APIs.

### Test Plan

```
$ rosetta-cli check:data --configuration-file rosetta_cli.json

[STATS] Blocks: 2600 (Orphaned: 0) Transactions: 5184 Operations: 36 Accounts: 5 Reconciliations: 49 (Inactive: 27, Exempt: 0, Skipped: 0, Coverage: 100.000000%)

Success: Reconciliation Coverage End Condition [Coverage: 100.000000%]

+--------------------+--------------------------------+--------+
|  CHECK:DATA TESTS  |          DESCRIPTION           | STATUS |
+--------------------+--------------------------------+--------+
| Request/Response   | Rosetta implementation         | PASSED |
|                    | serviced all requests          |        |
+--------------------+--------------------------------+--------+
| Response Assertion | All responses are correctly    | PASSED |
|                    | formatted                      |        |
+--------------------+--------------------------------+--------+
| Block Syncing      | Blocks are connected into a    | PASSED |
|                    | single canonical chain         |        |
+--------------------+--------------------------------+--------+
| Balance Tracking   | Account balances did not go    | PASSED |
|                    | negative                       |        |
+--------------------+--------------------------------+--------+
| Reconciliation     | No balance discrepancies were  | PASSED |
|                    | found between computed and     |        |
|                    | live balances                  |        |
+--------------------+--------------------------------+--------+

+--------------------------+--------------------------------+-------------+
|     CHECK:DATA STATS     |          DESCRIPTION           |    VALUE    |
+--------------------------+--------------------------------+-------------+
| Blocks                   | # of blocks synced             |        2600 |
+--------------------------+--------------------------------+-------------+
| Orphans                  | # of blocks orphaned           |           0 |
+--------------------------+--------------------------------+-------------+
| Transactions             | # of transaction processed     |        5184 |
+--------------------------+--------------------------------+-------------+
| Operations               | # of operations processed      |          36 |
+--------------------------+--------------------------------+-------------+
| Accounts                 | # of accounts seen             |           5 |
+--------------------------+--------------------------------+-------------+
| Active Reconciliations   | # of reconciliations performed |          22 |
|                          | after seeing an account in a   |             |
|                          | block                          |             |
+--------------------------+--------------------------------+-------------+
| Inactive Reconciliations | # of reconciliations performed |          27 |
|                          | on randomly selected accounts  |             |
+--------------------------+--------------------------------+-------------+
| Exempt Reconciliations   | # of reconciliation failures   |           0 |
|                          | considered exempt              |             |
+--------------------------+--------------------------------+-------------+
| Failed Reconciliations   | # of reconciliation failures   |           0 |
+--------------------------+--------------------------------+-------------+
| Skipped Reconciliations  | # of reconciliations skipped   |           0 |
+--------------------------+--------------------------------+-------------+
| Reconciliation Coverage  | % of accounts that have been   | 100.000000% |
|                          | reconciled                     |             |
+--------------------------+--------------------------------+-------------+

badger 2022/07/10 00:59:36 INFO: Storing value log head: {Fid:0 Len:32 Offset:3606533}
```

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/1867)

### CODE REVIEW

File: api/src/blocks.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        .with(metrics("get_genesis_ledger_info"))
+        .boxed()
+}
+
+async fn handle_get_block(
+    ledger_version: Option<LedgerVersionParam>,
+    context: Context,
+) -> Result<impl Reply, Rejection> {
+    fail_point("endpoint_get_block")?;
+    Ok(Block::new(ledger_version, context)?.account()?)
```
Comment (msmouse) [inline api/src/blocks.rs]: what is `account`?

File: api/src/blocks.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    ) -> Result<Self, Error> {
+        let latest_ledger_info = context.get_latest_ledger_info()?;
+        let ledger_version = ledger_version
+            .map(|v| v.parse("ledger version"))
+            .unwrap_or_else(|| Ok(latest_ledger_info.version()))?;
+
+        if ledger_version > latest_ledger_info.version() {
+            return Err(Error::not_found(
+                "ledger",
+                TransactionId::Version(ledger_version),
```
Comment (msmouse) [inline api/src/blocks.rs:46]: This will produce weird erro msg "ledger not found by ..."

File: api/src/blocks.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+        Ok(Self {
+            ledger_version,
+            latest_ledger_info,
+            context,
+        })
+    }
+
+    /// Scans the DB
+    pub fn find_block(self) -> Result<BlockDescription, Error> {
```
Comment (msmouse) [inline api/src/blocks.rs]: Make a API on DbReader similar to this, you will be able to get the block metadata txn version corresponding to an input version. (see how it calls the underlying `get_block_metadta()` which returns the version and the metadata

https://github.com/aptos-labs/aptos-core/blob/a87e1e0b448c6d884be7308f32fdc64a5a5ec360/storage/aptosdb/src/lib.rs#L1176

File: api/src/blocks.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                hash: hash.into(),
+            });
+        }
+
+        // Every other block needs to be found
+        const MAX_BLOCK_SIZE: u16 = 10000;
+        let search_start = self.ledger_version.saturating_sub(MAX_BLOCK_SIZE as u64);
+        let txns =
+            self.context
+                .get_transactions(search_start, MAX_BLOCK_SIZE * 2, self.ledger_version)?;
```
Comment (msmouse) [inline api/src/blocks.rs]: no no no, don't always load 20k txns just to find block boundary. call the db interface.

File: api/src/blocks.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            let txn_version = txn.version;
+            match &txn.transaction {
+                Transaction::BlockMetadata(_) => {
+                    // BlockMetadata is the beginning of the block
+                    if start < txn_version && txn_version <= self.ledger_version {
+                        start = txn_version;
+                    }
+                }
+                Transaction::StateCheckpoint => {
+                    // StateCheckpoint is at the end of the block
```
Comment (msmouse) [inline api/src/blocks.rs]: no, the state checkpoint is not a block epilog, don't assume it is a block boundary. We are ready to stop issuing state checkpoints as a way to boost performance in emergency.

File: api/src/context.rs [hunk 6]
```diff
@@ -90,6 +90,14 @@ impl Context {
         callback.await?
     }
 
+    pub fn get_genesis_accumulator_hash(&self) -> Result<HashValue, Error> {
+        let state_proof = self.db.get_state_proof_with_ledger_info(0)?;
+
+        Ok(state_proof
+            .latest_ledger_info()
+            .transaction_accumulator_hash())
+    }
```
Comment (msmouse) [inline api/src/context.rs]: We know this won't work after the pruner kicks in, right?

File: api/src/blocks.rs [hunk 7]
```diff
@@ -1,83 +1,31 @@
...
 use warp::{filters::BoxedFilter, Filter, Rejection, Reply};
 
-pub type BlockParam = Param<u64>;
-
-// GET /blocks/<address>
-pub fn get_block(context: Context) -> BoxedFilter<(impl Reply,)> {
-    warp::path!("blocks" / BlockParam)
+// GET /blocks/<version>
+pub fn get_block_info(context: Context) -> BoxedFilter<(impl Reply,)> {
+    warp::path!("blocks" / LedgerVersionParam)
```
Comment (msmouse) [inline api/src/blocks.rs:11]: nit: "ledger_version" in our context means you want to query things as if the entire ledger is frozen at "ledger_version". LedgerVersionParam is used as an optional query param in the API code. 

In your case you probably just want to do `path!("blocks" / Version)`


also "blocks/12345"  makes it look like you are asking for the 12345th block, which is not the case, right? (Although I don't have a better naming scheme in mind).

File: api/src/context.rs [hunk 8]
```diff
@@ -136,6 +130,109 @@ impl Context {
...
+        let converter = resolver.as_converter();
+        let txn = self.get_transaction_by_version(start, ledger_version)?;
+
+        // Parse the resources and find the block metadata resource update
+        let maybe_block_height = txn.changes.iter().find_map(|(key, op)| {
+            if let StateKey::AccessPath(path) = key {
+                if let Path::Resource(typ) = path.get_path() {
+                    // If it's block metadata, we can convert it to get the block height
+                    // And it must be the root address
+                    if path.address == aptos_root_address() && typ == block_metadata_type {
```
Comment (msmouse) [inline api/src/context.rs:200]: can't you just make a `StateKey` that represents `root_addr::[0x1::Block::BlockMetadta]` (and make it static), and search the change set by that?

File: api/src/context.rs [hunk 9]
```diff
@@ -136,6 +130,109 @@ impl Context {
...
+                        if let WriteOp::Value(value) = op {
+                            if let Ok(mut resource) = converter.try_into_resource(&typ, value) {
+                                if let Some(value) = resource.data.0.remove(height_id) {
+                                    if let Ok(height) = serde_json::from_value::<U64>(value) {
+                                        return Some(height.0);
+                                    }
+                                }
+                            }
```
Comment (msmouse) [inline api/src/context.rs:208]: Shall we define a Rust struct corresponding to the on chain struct and bsc deserialize the bytes into it? Using the converter will be slow.

Also @zekun000 , we can't make the height into the txn itself or the event?

File: storage/aptosdb/src/transaction_store/mod.rs [hunk 10]
```diff
@@ -150,6 +150,69 @@ impl TransactionStore {
...
+            let (v, txn) = res?;
+            // If we've found the beginning of the block, stop
+            if matches!(
+                txn,
+                Transaction::GenesisTransaction(_) | Transaction::BlockMetadata(_)
+            ) {
+                start_version = Some(v);
+                break;
+            }
+        }
```
Comment (msmouse) [inline storage/aptosdb/src/transaction_store/mod.rs:179]: Expecting genesis txn after genesis is kinda messy.. 

Shall we emit NewBlockEvent at both types of transactions and find block boundaries with that? @zekun000 
Also, can we assume the sequence number on the NewBlockEvent to be the block height? Or put the height in the event? @zekun000 
if the seq_num is the block height, we can get_block_by_block_index() using this index directly https://github.com/aptos-labs/aptos-core/blob/b1601d23fb157fee4c4932cd093d3809be8094a4/storage/aptosdb/src/schema/event_by_key/mod.rs#L10

we have this index https://github.com/aptos-labs/aptos-core/blob/b1601d23fb157fee4c4932cd093d3809be8094a4/storage/aptosdb/src/schema/event_by_version/mod.rs#L10 to find the sequence number for a specific event around a certain version directly, without a linear search. (get_block_metadata() can be updated to use it.)

File: storage/aptosdb/src/transaction_store/mod.rs [hunk 11]
```diff
@@ -150,6 +150,69 @@ impl TransactionStore {
...
+                Transaction::BlockMetadata(_) => {
+                    // If the current version is the beginning of the block, we need to find the
+                    // rest of the block
+                    if start_version != Some(v) {
+                        // This block metadata is the next block, so you need the previous version
+                        end_version = Some(v - 1);
+                        break;
+                    }
+                }
+                Transaction::StateCheckpoint => {
```
Comment (msmouse) [inline storage/aptosdb/src/transaction_store/mod.rs:196]: sorry but don't involve state checkpoint txns in deciding the block boundaries.

the logic can be:
```
let latset_ledger_version = get_latest_ledger_info().version();
// find the prev and next NewBlockEvent around version;
// if next new_block_event doesn't exist, but if the latest txn is at  latset_ledger_version, that's the end of block.
```

**Review by msmouse — APPROVED** [review]

**Review by msmouse — CHANGES_REQUESTED** [review]
(Sorry, approved accidentally)

File: api/src/context.rs (hunk 8 above)
Comment (gregnazario) [reply api/src/context.rs:200]: I'm a noob, and kinda been just doing it all on my own.  If that works, then I'll do that instead!

**Review by msmouse — APPROVED** [review]
unblock

File: api/src/context.rs (hunk 9 above)
Comment (zekun000) [reply api/src/context.rs:208]: we can add height to the event

**Pull request merged**